File: EduLink.Domain/Entities/Servicio.cs
using EduLink.Domain.Enums;
using EduLink.Domain.Interfaces;
using EduLink.Domain.Strategies;
using System;
using System.Collections.Generic;

namespace EduLink.Domain.Entities;

public class Servicio
{
    public Guid Id { get; init; } = Guid.NewGuid();
    public Guid ProveedorId { get; set; }
    public string Titulo { get; set; } = string.Empty;
    public string Descripcion { get; set; } = string.Empty;
    public decimal PrecioBase { get; set; }
    public int DuracionMinutos { get; set; }
    public Modalidad Modalidad { get; set; }
    public string? Ubicacion { get; set; }
    public List<SlotHorario> Slots { get; private set; } = new();

    public IPrecioStrategy EstrategiaPrecio { get; set; } = new PrecioBaseStrategy(); //elegir

    public decimal CalcularPrecioFinal(Cliente cliente) => EstrategiaPrecio.Calcular(PrecioBase, cliente); // strategi
}

File: app/domain/entities/servicio.rb
require 'securerandom'
require 'active_model'
require_relative '../strategies/precio_base_strategy'
require_relative '../entities/politica_cancelacion'

module Entities
  class Servicio
    include ActiveModel::Model

    attr_accessor :id, :proveedor_id, :titulo, :descripcion, :precio_base, :duracion_minutos, :modalidad, :ubicacion, :slots, :estrategia_precio, :politica_cancelacion

    def initialize(attributes = {})
      super
      @id ||= SecureRandom.uuid
      @slots ||= []

      # No es necesario definir un contexto porque ruby lo hace solito
      # Nuevamente se definen valores por defecto pero se pueden modificar con los getter/setter
      @estrategia_precio ||= Strategies::PrecioBaseStrategy.new

      @politica_cancelacion ||= PoliticaCancelacion.new(plazo_maximo_cancelacion_horas: 24, porcentaje_cargo: 0.1)
    end

    def calcular_precio_final(cliente)
      @estrategia_precio.calcular(@precio_base, cliente)
    end
  end
end
